proxy/grpcproxy/cache/store.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2016 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"hash/adler32"
+	"sync"
+
+	"fmt"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/mvcc"
+	"github.com/golang/groupcache/lru"
+)
+
+const DefaultMaxEntries = 2048
+
+type Cache interface {
+	Add(req *pb.RangeRequest, resp *pb.RangeResponse)
+	Get(req *pb.RangeRequest) (*pb.RangeResponse, bool, error)
+	Compact(revision int64)
+}
+
+// keyFunc returns the key of an request, which is used to look up in the cache for it's caching response.
+func keyFunc(req *pb.RangeRequest) uint64 {
+	hash := adler32.New()
+	fmt.Fprintf(hash, "%#v", req)
+	return uint64(hash.Sum32())
+}
+
+func NewCache(maxCacheEntries int) Cache {
+	return &cache{
+		lru: lru.New(maxCacheEntries),
+	}
+}
+
+// cache implements Cache
+type cache struct {
+	lock                sync.RWMutex
+	lru                 *lru.Cache
+	latestCompactionRev int64
+}
+
+// Add adds response for a request to the cache.
```

Comment:
Add adds the response of a request to the cache if its revision is larger than the compacted revision of the cache.